src/api.rs

```diff
@@ -0,0 +1,90 @@
+//! This file contains the API for the runtime GC. Interaction with the GC is
+//! only legal through these functions. 
+//!
+//! The GC runtime consists of a mutable singleton `Collector` struct: 
+//!
+//! ```rust, ignore
+//! struct Collector { 
+//!     // Heap bookkeeping information
+//!     hstart: *const usize,
+//!     hptr: *mut usize,
+//!
+//!     // Flag to determine whether to collect at the next safepoint
+//!     collect_next: bool,
+//!
+//!     // The in-memory safepoint table used to identify roots
+//!     roots: HashMap<ReturnAddress, SafepointRoots>
+//! }
+
+use collector::COLLECTOR;
+
+/// This must be called before the GC can be used (usually in the setup code
+/// before `main()`). Initialisation consists of two stages:
+///     1. Read the stackmap section in the ELF file into an in-memory table for
+///        fast lookup.
+///     2. Allocate a chunk of heap memory to be used to store objects managed
+///        by the GC.
+pub fn init() {
+        COLLECTOR.alloc_heap().make_root_table();
+}
+
+/// This function is the *only* way that a collection can be triggered. Calls to
+/// `safepoint_poll` are generated by LLVM's InsertSafepoints opt pass. They are
+/// inserted liberally into the mutator's code at all function calls and
+/// loop-backedges. 
+///
+/// In almost all cases, a safepoint poll will *never* trigger a collection.
+/// It's therefore really important that this function returns fast on a
+/// wont-collect poll.
+/// 
+/// No information about whether a poll resulted in a collection is returned to
+/// the mutator. The only thing that can be guaranteed is that a collection
+/// *might* have happened after returning from this call.
+///
+/// ------------------------------ WARNING -------------------------------------
+/// | We need to be really careful about code which panics inside the collector|
+/// | here. The safepoint poll will *not* be called by native Rust code which  |
+/// | means that panic handling is UB. We should probably abort on panic here. |
+/// ----------------------------------------------------------------------------
```

Comment:
Should we use the `abort` panic strategy then?

Reply:
This is tricky: as a library, our GC can't really impose panic or unwind on users. Ideally, we have to deal with whatever they select. I agree that unwind feels a bit of a mess in Rust.

Reply:
I'm not sure I have a good answer to this one. At the very least, when debugging, we'll want abort on panic so that we can have some idea what went wrong. But yes, for the user we would need to hopefully recover in some way.

Reply:
Can the user really recover if the GC fails?

Reply:
Yes. You can sometimes free up inessential resources (e.g. caches). It doesn't happen often though. Anyway, that's an orthogonal issue: in the long run we *have* to support unwinding, even if we don't like it very much.

Reply:
I'm sceptical, but let's see what happens.